ui/apps/dashboard/src/components/Insights/InsightsChat/InsightsChat.tsx

```diff
@@ -0,0 +1,296 @@
+                ? (data as { metadata?: { toolName?: string } }).metadata!.toolName
+                : undefined;
+            const partId = typeof data['partId'] === 'string' ? data['partId'] : undefined;
+            if (type === 'tool-call') setCurrentToolName(tn || null);
+            if ((type === 'tool-output' || type === 'tool-call') && partId && tn) {
+              partIdToToolNameRef.current.set(partId, tn);
+            }
+            break;
+          }
+          case 'part.completed': {
+            const type = typeof data['type'] === 'string' ? data['type'] : undefined;
+            const partId = typeof data['partId'] === 'string' ? data['partId'] : undefined;
+            if (type === 'text') {
+              setTextStreaming(false);
+              setTextCompleted(true);
+            } else if (type === 'tool-output' || type === 'tool-call') {
+              setCurrentToolName(null);
+              // Auto-paste and run SQL when generate_sql tool output completes
+              if (type === 'tool-output' && partId && !autoRanPartIdsRef.current.has(partId)) {
+                const toolName = partIdToToolNameRef.current.get(partId);
+                if (toolName === 'generate_sql') {
+                  autoRanPartIdsRef.current.add(partId);
+                  const finalContent = (data as { finalContent?: unknown }).finalContent;
+                  let sql: string | null = null;
+                  try {
+                    if (finalContent && typeof finalContent === 'object') {
+                      const obj = finalContent as Record<string, unknown>;
+                      const envelope = (obj as { data?: unknown }).data as
+                        | Record<string, unknown>
+                        | undefined;
+                      const candidate = envelope?.sql ?? (obj as { sql?: unknown }).sql;
+                      if (typeof candidate === 'string' && candidate.trim()) sql = candidate.trim();
+                    } else if (typeof finalContent === 'string') {
+                      try {
+                        const parsed = JSON.parse(finalContent) as Record<string, unknown>;
+                        const envelope = (parsed as { data?: unknown }).data as
+                          | Record<string, unknown>
+                          | undefined;
+                        const candidate = envelope?.sql ?? (parsed as { sql?: unknown }).sql;
+                        if (typeof candidate === 'string' && candidate.trim())
+                          sql = candidate.trim();
+                      } catch {
+                        // Not JSON; ignore
+                      }
+                    }
+                  } catch {}
+                  if (sql) {
+                    try {
+                      onSqlChange(sql);
+                      runQuery();
+                    } catch {}
+                  }
+                }
+              }
+            }
+            break;
```

Comment:
Is there a way to simplify this block of code? Or at least add comments to help readers understand what's happening? There's also tons of type casting in here